File: app/src/main/java/crux/bphc/cms/app/MyApplication.java
package crux.bphc.cms.app;

import android.app.Application;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.Gson;

import java.util.HashMap;

import io.realm.Realm;
import io.realm.RealmConfiguration;

public class MyApplication extends Application {

    private static MyApplication mInstance;
    private boolean isDarkMode = false;

    public static synchronized MyApplication getInstance() {
        return mInstance;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        mInstance = this;

        initRealm();

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        this.isDarkMode = prefs.getBoolean(Constants.DARK_MODE_KEY,false);
    }

    public boolean isDarkModeEnabled() {
        return this.isDarkMode;
    }

    public void setDarkModeEnabled(boolean isEnabled) {
        this.isDarkMode = isEnabled;
        SharedPreferences.Editor prefEdit = PreferenceManager.getDefaultSharedPreferences(this).edit();
        prefEdit.putBoolean(Constants.DARK_MODE_KEY, this.isDarkMode);
        prefEdit.apply();
    }

    @SuppressWarnings("unchecked")
    public HashMap<String, String> getLoginLaunchData() {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        String data = prefs.getString(Constants.LOGIN_LAUNCH_DATA, "");
        return new Gson().fromJson(data, HashMap.class);
    }

    public void setLoginLaunchData(HashMap<String, String> launchData) {
        SharedPreferences.Editor prefEdit = PreferenceManager.getDefaultSharedPreferences(this).edit();
        prefEdit.putString(Constants.LOGIN_LAUNCH_DATA, new Gson().toJson(launchData));
        prefEdit.apply();
    }

    private void initRealm() {
        Realm.init(this);
        Realm.setDefaultConfiguration(getRealmConfiguration());
    }

    private RealmConfiguration getRealmConfiguration() {
        return new RealmConfiguration.Builder()
                .deleteRealmIfMigrationNeeded()
                .build();
    }
}

File: app/src/main/java/crux/bphc/cms/app/MyApplication.kt
package crux.bphc.cms.app

import android.app.Application
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import io.realm.Realm
import io.realm.RealmConfiguration

class MyApplication : Application() {
    private lateinit var sharedPref: SharedPreferences

    private val realmConfiguration: RealmConfiguration
        get() = RealmConfiguration.Builder()
            .deleteRealmIfMigrationNeeded()
            .build()

    val loginLaunchData: HashMap<String, String>
        get() {
            val prefs = PreferenceManager.getDefaultSharedPreferences(this)
            val data = prefs.getString(Constants.LOGIN_LAUNCH_DATA, "")
            val type = object : TypeToken<HashMap<String, String>>(){}.type
            return Gson().fromJson(data, type)
        }

    override fun onCreate() {
        super.onCreate()
        instance = this
        initRealm()
        sharedPref = PreferenceManager.getDefaultSharedPreferences(this)
    }

    private fun initRealm() {
        Realm.init(this)
        Realm.setDefaultConfiguration(realmConfiguration)
    }

    fun setLoginLaunchData(loginLaunchData: HashMap<String, String>) {
        sharedPref.edit()
            .putString(Constants.LOGIN_LAUNCH_DATA, Gson().toJson(loginLaunchData))
            .apply()
    }

    companion object {
        @get:Synchronized
        lateinit var instance: MyApplication
            private set
    }
}